### ISSUE TITLE
make docs-all error

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1876]
@Bultako - I'm getting this error running `make docs-all`:
https://gist.github.com/cdeil/eee297e2d6c32a60d16dcde1fd610156

I think what happens is that this is called:
https://github.com/gammapy/gammapy/blob/b17e6b1d20ae9cbe64116d5656417bfbf6db2564/Makefile#L103
which makes a subprocess call to `gammapy`:
https://github.com/gammapy/gammapy/blob/b17e6b1d20ae9cbe64116d5656417bfbf6db2564/gammapy/utils/tutorials_process.py#L90
which in my case  is pointing to `/Users/deil/work/code/gammapy-docs/build/dev/gammapy` where I did my docs build, because there I also ran `make docs-all` which called `pip install -e .` whic put a link in my site-packages which now wins for some reason. So some other Gammapy version is executed, not the one I wanted.

I think if we don't shell out via subprocess, and instead import and call the click CLI directly, we guarantee that the "right" Gammapy is used and the issue is resolved. I could find an example how to do that or send a PR if that's the say to go.

@Bultako - Thoughts?

**Comment by cdeil** [comment]
A similar (but different) issue might occur here:

https://github.com/gammapy/gammapy/blob/b17e6b1d20ae9cbe64116d5656417bfbf6db2564/gammapy/utils/tutorials_process.py#L107-L109

If the PATH isn't nice for some reason, that could call the "wrong" jupyter from some other Python.
Putting `cmd = [sys.executable, '-m', 'jupyter']` instead of just `jupyter` is a little bit safer, because it guarantees that the "right" Python is used.
Not sure if this is a real issue and whether it's worth making the change. I just mention it because I saw the other subprocess call in that file.

**Comment by Bultako** [comment]
Ok
I understand..

I could do a PR modifying **all** the `subprocess.call` syntax found in the code, and replacing the calling with a string by calling with a list of params (with sys.executable)

i.e. 

```python
subprocess.call("gammapy jupyter --src temp black", shell=True)
```

by 

```
subprocess.call([sys.executable, '-m', 'gammapy', 'jupyter --src tempblack'])
```

Do you think this could work?

**Comment by cdeil** [comment]
Yes, that sounds good. Thank you!

**Comment by cdeil** [comment]
Resolved by @Bultako  in #1879

**Comment by cdeil** [comment]
Actually, I'm still getting this error, so am not able to bulid the devdocs locally at the moment and resolve #1875.

This is what we do at the moment:

https://github.com/gammapy/gammapy/blob/2490b4671c281b261b7b237f939220b789cba5ba/gammapy/scripts/jupyter.py#L48-L60

We do a subprocess, and execute the right Python & Jupyter there (the one from the currently active env, which for me is gammapy-dev. But then Jupyter apparently starts a Python kernel, which doesn't load the `gammapy` from the local folder where I execute the command (`/Users/deil/work/code/gammapy-docs/build/dev/gammapy`, which I build the docs), but from the other `/Users/deil/work/code/gammapy/gammapy` which is my development version, which isn't built at the moment or might have another branch or modifications).

@Bultako - Is it clear? Can you reproduce? Any idea how to get Jupyter to import the local gammapy, e.g. by passing some setting or env variable or sys.path manipulation?

For now, I'll just manually make sure I have the latest dev docs version over at `/Users/deil/work/code/gammapy/gammapy` also and do a build and push the HTML once.

**Comment by Bultako** [comment]
I guess we could pass the python kernel as a param in the `make docs-all` command.
In the case you had problems with this, you could do the following workflow:

```
$ jupyter kernelspec list
Available kernels:
  cta-dev        /Users/jer/Library/Jupyter/kernels/cta-dev
  gammapy-dev    /Users/jer/Library/Jupyter/kernels/gammapy-dev
  joint-crab     /Users/jer/Library/Jupyter/kernels/joint-crab
  python3        /Users/jer/anaconda/envs/joint-crab/share/jupyter/kernels/python3
```

and then pass the kernel name that you choose to `make docs-all`

```
make docs-all kernel=gammapy-dev
```

@cdeil 
Do you already get the kernel name you want with `jupyter kernelspec list`?
In that case, I could do that modif in the scripts.

**Comment by cdeil** [comment]
I get this when I have the `gammapy-dev` env active:
```
(gammapy-dev) hfm-1804a:gammapy deil$ jupyter kernelspec list
Available kernels:
  python3    /Users/deil/software/anaconda3/envs/gammapy-dev/share/jupyter/kernels/python3
  xonsh      /Users/deil/software/anaconda3/envs/gammapy-dev/share/jupyter/kernels/xonsh
```

I don't think some kernel selection will here here?

**Comment by Bultako** [comment]
The kernel used in `make docs-all` is `python3` 
If you type for the `python3` kernel
```
cat /Users/deil/software/anaconda3/envs/gammapy-dev/share/jupyter/kernels/python3/kernel.json
```
you will see where exactly this kernel is pointing.
In my case, I got this output.

```
{
 "argv": [
  "/Users/jer/anaconda/envs/gammapy-dev/bin/python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```

If your `python3` kernel path is pointing to your local git repo of the devel version of `gammapy`, maybe you should just create a branch in this repo when you try to build the docs, and run the `make docs-all` there. I think this workflow could solve the issue.

**Comment by cdeil** [comment]
I have the same `kernel.json` as you.

I think what's happening is that Jupyter puts the local working directory last on `sys.path`. That's why the local verion of Gammapy in `/Users/deil/work/code/gammapy-docs/build/dev/gammapy` isn't picked up, only a version from `/Users/deil/software/anaconda3/envs/gammapy-dev/lib/python3.6/site-packages`:
```
$ jupyter troubleshoot
<snip/>
sys.path:
	/Users/deil/software/anaconda3/envs/gammapy-dev/bin
	/Users/deil/software/anaconda3/envs/gammapy-dev/lib/python36.zip
	/Users/deil/software/anaconda3/envs/gammapy-dev/lib/python3.6
	/Users/deil/software/anaconda3/envs/gammapy-dev/lib/python3.6/lib-dynload
	/Users/deil/software/anaconda3/envs/gammapy-dev/lib/python3.6/site-packages
	/Users/deil/work/code/gammapy-docs/build/dev/gammapy
<snip/>
```
I'm not sure why the `pip install -e .` from the `Makefile` doesn't ensure that the right local version is installed. I guess there could be an issue with running editable installs for the same package from two locations.

Overall I'm not sure what to do here. I mean: I can get it to work, it's fine for now. But the setup is error-prone, so I think future maintainers of the docs build might run into the same issue.
Since you don't see the issue, I could either investigate a bit more and try some things, e.g. to change to `pip install .` for `make docs-all`, or we could try Docker. But I guess with Docker it might just be a different can of worms?

**Comment by Bultako** [comment]
I have tried with `pip install .` in the Makefile, and it continues placing the last installed path at the end of the list in `sys.path`. But,  I have tried adding to the Makefile `pip uninstall gammapy` before `pip install .` and it seems to work.

I would prefer  to keep this `pip uninstall gammapy` outside the Makefile. I think this is more related with a very specific maintainer's environment or work process. I would be in favor of standardizing and documenting a doc-maintenance workflow, from what already exists in
https://github.com/gammapy/gammapy-docs

From what you say, it seems you're working with two different local repos of the same `gammapy` Github repository, so you have one local repo to work with the code and a second one to maintain the docs, and both activities need a different evolving version of gammapy of the same repo.  

Should we think of a workflow using two different conda envs, or different local git branches in a single local repo? Or adding `pip uninstall gammapy` to the recipe, and then again re-install the version of **code repo** when coming back to work with this one? 

Let's give it a thought on how it is best for you (doc and code maintainers) to switch from one gammapy env to another.

**Comment by cdeil** [comment]
The workflow I use is to run the docs build in a sub-folder of the `gammapy-docs` repo, as documented here:
https://github.com/gammapy/gammapy-docs#update-dev-version
This is where the issue occurs.

And yes, I have a separate git clone of the Gammapy repo where I do my coding.

I'm also -1 to uninstall something from the Makefile.

> Should we think of a workflow using two different conda envs, 

If we don't find a solution to use the existing `gammapy-dev` environment, and get a consistent docs build using that in the `gammapy-docs` repo, then yes, maybe creating a dedicated env for the docs build might be a good idea.

> or different local git branches in a single local repo?

Without having thought about it much, I think I'm -1.

The docs build is very slow, I prefer to have it in a separate repo and not be blocked to work on the Gammapy code by being stuck in some branch for 10 min.

> Or adding pip uninstall gammapy to the recipe, and then again re-install the version of the gammapy code repo when coming back to work with this one?

Again, without having thought about it much, I think I'm -1.

Basically I don't want my `gammapy-dev` env affected by running a docs build in `gammapy-docs`.
That would mean that there we should not run `pip install -e .`, but rather `python setup.py build_ext -i`. Then the site-packages will not be altered.

**Comment by Bultako** [comment]
> Basically I don't want my gammapy-dev env affected by running a docs build in gammapy-docs.
That would mean that there we should not run `pip install -e .`, but rather `python setup.py build_ext -i`. Then the site-packages will not be altered.

In the docs-building process, we need a kernel pointing to a full valid install of gammapy for `jupyter nbconvert` to execute/fill and test the notebooks.

If we do want to keep two different versions of gammapy, one for docs and one for code, I would go for virtual envs and clone to a `gammapy-docs` conda env.

```
conda create --name gammapy-docs --clone gammapy-dev
pip uninstall gammapy
```

And then switching to and from it when building the docs.

This would mean these two envs could also dverge with time because of the different versions of the packages installed.

**Comment by cdeil** [comment]
I did change the instructions to use `gammapy setup.py develop`, to avoid modifying site-packages. I think with this the docs build worked fine for me. Although it could be that it sometimes works fine, and sometimes not, namely when there's a link in site-packages from a previous `pip install -e .` insteall of Gammapy from the code / dev repo.

https://github.com/gammapy/gammapy-docs/blob/master/README.md#update-dev-version

So I'm keeping this issue open for now and assigning to myself. But possibly there's nothing to do apart from improving the maintainer docs. I will close it once the dev docs build runs on Azure pipelines and works fine (which is what the remainig task in #1241 is).

**Comment by cdeil** [comment]
I just now did a dev docs build locally following the instructions at https://github.com/gammapy/gammapy-docs#update-dev-version .

And we have dev docs builds in CI that are working, e.g. here: https://github.com/gammapy/gammapy/blob/0b100d3abd3a741871b7844925f46ffe08183cc8/azure-pipelines.yml#L68

So I'm closing this old issue.

Future Gammapy maintainers: if you find a similar issue in the future, where the wrong Python is picked up for parts of the docs build execution, please open a new issue and if useful reference this one.
